만국박람회 [STEP2] Jeremy, 준호

yjjem opened the pull request:
안녕하세요 ***`올라프`*** ☃️
@1Consumption
Jeremy 와 준호 입니다.

## ✏️ 구현 내용
* 첫 번째 화면은 스크롤 뷰와 스택 뷰를 활용해서 만들었습니다.
* 두 번째 화면으로 이동은 segue를 활용했습니다.
* 두 번째 화면은 테이블 뷰를 활용해서 만들었습니다.
* 세 번째 화면으로 이동은 NavigationController를 활용했습니다.
* 세 번째 화면은 스택 뷰를 활용해서 만들었습니다

<center>
    <div style="display: flex; flex-direction: row; justify-cotents: center;">
        <img src="https://i.imgur.com/ja53aMa.jpg" width="200px"/>
        <img src="https://i.imgur.com/668cK11.jpg" width="200px"/>
        <img src="https://i.imgur.com/reqRHGC.png" width="200px"/>
    </div>
</center>

## 코드를 작성할 때 고민되었던 점
* 첫 번째 화면과 두 번째 화면에서 JSONDecoder를 사용하는 코드가 비슷했습니다. 중복을 줄이기 위해서 JSONDecoder을 extension하여 decode 메서드를 추가했습니다. 제네릭 타입을 사용해서 두 화면 모두에서 사용할 수 있도록 했습니다.
    ```swift
    extension JSONDecoder {

        func decode<T: Decodable>(from fileName: String, to: T) -> T? {
            guard let dataAsset: NSDataAsset = NSDataAsset(name: fileName) else { return nil }
            do {
                return try self.decode(T.self, from: dataAsset.data)
            } catch {
                return nil
            }
        }
    }
    ```
* Custom Cell을 사용하지 않고 기본 셀로 구현중에 왼쪽 그림과 같이 imageView에 constraints가 잡히지 않는 점이 고민이였습니다. AutoLayout 관련 문제들은 Step3에서 진행한다고 해서 아직 깊게 고민하지는 않았습니다. 임시방편으로 가독성을 위해 heightForRowAt을 사용하여 오른쪽의 그림과 같이 cell의 크기를 잡아주었습니다.

<div style="display: flex; flex-direction: row;">
    <img src="https://i.imgur.com/g3445A1.jpg" width="300px"/>
    <img src="https://i.imgur.com/J771rfn.jpg" width="300px"/>
</div>

    

## 조언을 얻고 싶은 부분

* 예시처럼 첫 번째 화면은 네비게이션 바를 가리기 위해서 아래와 같이 코드를 추가했습니다.
* 첫 번째 화면의 viewWillAppear 메서드에서 navigationBar를 숨기고, 두 번째 화면의 viewWillAppear 메서드에서 navigationBar를 표시했습니다.
* 더 좋은 방법이 있는지 궁금합니다.


<div style="display: flex; flex-direction: row;">
    <div style="background-color: black; width: 35%"></div>
    <img src="https://i.imgur.com/ky1xhMT.png" width="300px"/>
    <div style="background-color: black; width: 35%"></div>
</div>

```swift
    // 첫 번째 화면
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        self.navigationController?.isNavigationBarHidden = true
    }

    // 두 번쨰 화면
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        self.navigationController?.isNavigationBarHidden = false
```

junho15 committed:
feat: 첫화면 storyboard 구현

yjjem committed:
feat: 포스터와 각종 설명 표시 기능 추가

junho15 committed:
feat: storyBoard 구현 (view, tableView), ExpoEntryViewController 추가, viewController 이름을 ExpositionViewController로 수정

yjjem committed:
feat: 데이터 추출 기능 추가, TableView 기능 구현

junho15 committed:
feat: 세번째 화면 storyBoard 구현 (view, tableView), EntryViewController 추가, ExpoEntryViewController 이름을 ExpoEntriesViewController로 수정

yjjem committed:
feat: segue 대신 NavigationController를 사용하여  화면 이동, 이미지 및 텍스트 표시 기능 추가

junho15 committed:
refactor: 첫 화면 방문객, 개최지, 개최기간 Label 분리

yjjem committed:
feat: JSONDecoder extenstion을 추가하여 기존 코드 대체, Navigation Bar 표시 제어기능 추가

junho15 committed:
refactor: Model, Controller, View 폴더별로 구분

yjjem committed:
refactor: 켄벤션 및 네이밍 수정

junho15 committed:
refactor: JSONDecoder 파일명을 JSONDecoder+decode로 변경

yjjem committed:
refactor: scrollView AutoLayout 수정

yjjem committed:
Update README.md

1Consumption commented on Expo1900/Expo1900/Controller/EntryViewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+import UIKit
+
+final class EntryViewController: UIViewController {
```
상세화면을 보여주는 곳이니까 이름에 detail이 들어가면 어떨까요? 이름만 보고는 어떤 역할을 하는지 유추하기 힘드네요.

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/pull/219/commits/d08d7fe74ad6887fc22ae060bf1252870ca23c1b)

확실히 한눈에 알아보기 힘들게 네이밍을 했군요. 말씀해주신 대로 detail을 넣어주면 좋을 것 같습니다. 아래와 같이 수정합니다.
```swift
final class EntryDetailViewController: UIViewController { ... }
```

1Consumption commented on Expo1900/Expo1900/Controller/EntryViewController.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    private var expoEntry: ExpoEntry?
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        
+        configureEntryViews(expoEntry: expoEntry)
+    }
```
내부 프로퍼티에 값을 할당해주는 코드를 viewWillAppear에서 부르는 이유가 있나요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/pull/219/commits/dd881f13adaf1813e300de61f0f775d0988bf90e)

세 번째 뷰가 다시 열릴 때 disappear 됐다가 appear 된다고 착각을 했습니다. viewDidLoad()에서 호출하도록 수정했습니다.

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    @IBOutlet private weak var ExpoEntriesTableView: UITableView!
+    
+    var expoEntries: [ExpoEntry] = []
```
internal인 이유가 있나요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/pull/219/commits/285503e0e87e4722be3275a41978b42a5721dd74)

해당 변수는 클래스 내에서만 사용되어야 합니다. 따라서 itnernal일 이유가 없고 private지정을 해주어야 한다고 생각합니다.
expoEntries를 expoEntriesManger로 변경하면서 사라진 내용이지만 같은 맥락에서 expoEntrieManager를 private 지정해주었습니다. 감사합니다.
```swift
private var expoEntriesManager: ExpoEntriesManager = ExpoEntriesManager()
```

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+        super.viewWillAppear(animated)
+        
+        self.navigationController?.isNavigationBarHidden = false
+        configureExpoEntriesFromJSON(name: "items")
```
제이슨 디코딩 하는 코드도 viewWillAppear에 있는 이유가 있나요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/285503e0e87e4722be3275a41978b42a5721dd74)

viewDidLoad()에서 expoEntriesManager의 `configureExpoEntriesFromJSON` 메서드를 호출하도록 수정했습니다. viewWillAppear()에 추가하면 세 번째 화면에서 두 번째 화면으로 돌아올 때 다시 호출되는 문제가 있어서 viewDidLoad로 옮겼습니다.

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+        configureExpoEntriesFromJSON(name: "items")
+    }
+    
+    private func configureExpoEntriesFromJSON(name: String) {
+        guard let expoEntries =  JSONDecoder().decode(from: name, to: expoEntries) else { return }
+        self.expoEntries = expoEntries
+    }
```
이런 코드는 View에 있는게 맞을까요? View에서 어떤 이름의 asset을 디코딩하는지까지 알아야하나요?
이런 모든 역할을 추상화한 객체를 View는 가지고만있고, 이벤트를 보내면 어떨까요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/285503e0e87e4722be3275a41978b42a5721dd74)

말씀해주신대로 View가 Model의 일을 하는 것 같습니다. ExpoEntriesManager 구조체를 추가하여 기능을 분리했습니다. ExpoEntriesManager의 `configureExpoEntriesFromJSON` 메서드를 호출해서 디코딩하도록 했습니다.

---
❓ 이벤트를 보낸다는게 정확히 어떤 것을 의미하는지 잘 모르겠습니다.

1Consumption replied:
지금처럼 view에서 manager의 메소드를 호출하는걸 이벤트를 전달한다고 해요

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+
+final class ExpoEntriesViewController: UIViewController {
+    
+    @IBOutlet private weak var ExpoEntriesTableView: UITableView!
```
대문자인 이유가 있나요?

yjjem replied:
오타 수정했습니다. 감사합니다.

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: UITableViewCell = ExpoEntriesTableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
```
매개변수로 주어진 tableView를 사용하지 않은 이유가 있을까요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/pull/219/commits/285503e0e87e4722be3275a41978b42a5721dd74)

음.. 사실 생각없이 지나치던 부분 같습니다. 
확인해보니 ViewController에 연결된 TableView를 파라매터로 받고 있는 것을 확인하여 직접적으로 `expoEntriesListTableView`에 접근하지 않고 매개변수의 tableview를 활용했습니다.
![](https://i.imgur.com/bc4DOsd.png)

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+        return cell
+    }
+    
+    private func configureExpoEntry(of cell: UITableViewCell, with expoEntry: ExpoEntry) {
+        cell.imageView?.image = UIImage(named: expoEntry.imageName)
+        cell.textLabel?.text = expoEntry.name
+        cell.detailTextLabel?.text = expoEntry.shortDescription
+    }
```
이 코드는 ViewController에 있는게 아니라 Cell에 있어야할 코드 같습니다. 지금은 Cell이 별로 없지만 만약 Cell이 정말 많아졌을 때 그 모든 코드가 ViewController에 있다면 어떨까요? 그리고 지금은 별 로직이 없지만 복잡한 로직이 들어간다면 그것 또한 ViewController에 들어가야할까요?

yjjem replied:
오타 수정했습니다. 감사합니다.

1Consumption replied:
오타수정이요?

yjjem replied:
위의 코멘트는 정정합니다.. 혼란을 드려 죄송합니다

[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/285503e0e87e4722be3275a41978b42a5721dd74)

ExpoEntryCell 커스텀 셀을 만들어 configureContentsView 메서드로 cell의 contentView의 imageView와 label을 설정하도록 했습니다
    
```swift
// ExpoEntryCell.swift
class ExpoEntryCell: UITableViewCell {
    func configureContentsView(image: UIImage?, name: String, shortDescription: String) {
        self.imageView?.image = image
        self.textLabel?.text = name
        self.detailTextLabel?.text = shortDescription
    }
}

// ExpoEntriesListViewController.swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell 
    guard let cell: ExpoEntryCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indxPath) as? ExpoEntryCell,
          let expoEntry = expoEntriesManager.expoEntries?[indexPath.row] else { return ExpEntryCell() }
    let image: UIImage? = expoEntriesManager.expoEntryImage(index: indexPath.row)
    cell.configureContentsView(image: image, name: expoEntry.name, shortDescription: expEntry.shortDescription)
    return cell
}
```

1Consumption commented on Expo1900/Expo1900/Controller/ExpoEntriesViewController.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let entryViewController: EntryViewController = self.storyboard?.instantiateViewController(withIdentifier: "EntryViewController") as? EntryViewController  else { return }
+        entryViewController.sendExpoEntry(expoEntries[indexPath.row])
```
데이터를 전달해주는 코드 같습니다.
[instantiateViewController(identifier:creator:)](https://developer.apple.com/documentation/uikit/uistoryboard/3213989-instantiateviewcontroller) 이 메소드의 creator 매개변수는 NSCoder를 매개변수로 전달하고 ViewController를 반환하는 클로저인데, 이를 통해 NSCoder를 가져올 수 있습니다.

그래서 ViewController에 custom initializer를 생성해 아래와 같이 init시점에 데이터를 전달할 수도 있습니다.

```
class ViewController {

    init?(coder: NSCoder, entry: ExpoEntry) {
        super.init(coder: coder)
        self.entry = entry
    }
}
```

추가적으로 `init(coder:)` 와 `init(frame:)`의 차이는 뭘까요?

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/dd881f13adaf1813e300de61f0f775d0988bf90e)

`init(coder:)` 스토리보드에 만들어진 XML형태의 xib파일을 컴파일하여 바이너리 nib 파일로 unarchiving하여 자신을 반환
`init(frame:)` 코드로 View를 생성할 떄 CGRect(x:y:width:height:) 를 사용하여 위치와 크기를 설정

ViewController에 init?()을 추가하고 instantiateViewController(identifier:creator:) 를 사용하도록 수정했습니다. init?()을 선언하기 위해서 필요한 required init?()도 추가했습니다.
```swift
// EntryDetailViewController.swift
    init?(coder: NSCoder, entry: ExpoEntry) {
        super.init(coder: coder)
        self.expoEntry = entry
    }
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }    

// ExpoEntriesListViewController.swift
    guard let entryDetailViewController: EntryDetailViewController = sel.storyboard?.instantiateViewController(identifier: "EntryViewController", creator: { (coder) -> EntyDetailViewController? in
        return EntryDetailViewController(coder: coder, entry: expoEntry)
    }) as? EntryDetailViewController else { return }
    self.navigationController?.pushViewController(entryDetailViewController, animated: true)
``

1Consumption commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,48 @@
+        durationLabel.text = exposition?.duration
+        descriptionTextView.text = exposition?.description
+    }
+    
```
불필요한 공백 같습니다.

yjjem replied:
수정했습니다. 갑사합니다.

1Consumption commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,48 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        configureExpositionFromJSON(name: "exposition_universelle_1900")
```
asset 이름을 아는건 View에서 할 일이 아니라고 생각합니다.

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/7d4be9805372f8c7ec88f0dc4a755b84449fbba9)

ExpositionManager의 posterImage 연산 프로퍼티를 사용해서 asset의 이미지를 반환하도록 했습니다.
```swift
    var posterImage: UIImage? {
        return UIImage(named: imageName)
    }
```

1Consumption replied:
지금은 에셋이 로컬에 있어서 항상 로컬에 있는 이미지를 받아오니까 이런 코드를 사용할 수 있는데, 만약 서버에서 이미지를 받아온다면 지금과 같은 코드를 사용할 수 있을까요?

1Consumption commented on Expo1900/Expo1900/Model/Exposition.swift:
```diff
@@ -0,0 +1,22 @@
+    let duration: String
+    let description: String
+    
+    var visitorsDescription: String {
+        let numberFormatter: NumberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        let visitorsDescription: String = numberFormatter.string(from: NSNumber(value: visitors)) ?? ""
+        return visitorsDescription + " 명"
+    }
```
음... 이게 여기서 할 일인지는 잘 모르겠어요. 그냥 데이터를 담는 컨테이너라고 생각했어요. 그래서 따로 이런 로직을 가지고 모델을 관리하는 객체가 있으면 좋겠습니다.

yjjem replied:
[`📝 수정 내용`](https://github.com/yagom-academy/ios-exposition-universelle/commit/7d4be9805372f8c7ec88f0dc4a755b84449fbba9)

ExpositionManager 구조체를 추가했습니다. ExpositionManager의 formattedVisitorsDescription 연산 프로퍼티로 처리하도록 했습니다.
```swift
var formattedVisitorsDescription: String? {
    let numberFormatter: NumberFormatter = NumberFormatter()
    numberFormatter.numberStyle = .decimal
    guard let visitors = self.exposition?.visitors,
          let visitorsDescription = numberFormatter.string(from: NSNumber(value: visitors)) els { return nil }
    return visitorsDescription + " 명"
    }
```

yjjem replied:
말씀해주신 대로 데이터를 담는 컨테이너의 역할만 하도록 수정을 했습니다. 이 과정에서 궁금한 점이 있습니다. 
관리하는 객체가 있으면 좋겠다고 말씀을 해주셨는데 extension으로 빼는 것은 기능 분리와는 조금 다른 개념이 되는건가요?

1Consumption replied:
extension은 말 그대로 해당 객체에 확장을 통해 기능을 추가하는거고, 관리하는 객체를 만드는건 새로운 객체를 만드는거죠.
객체의 역할을 분리한다는 점에서 다르다고 생각합니다

1Consumption reviewed (changes requested):
안녕하세요 @yjjem @junho15 고생 많으셨습니다~

구현 내용을 적어주시기 보다는 왜 그렇게 구현했는지를 적어주시면 좋겠습니다.
이번 스텝 또한 별다른 코멘트는 없습니다.

조언을 얻고싶은 부분에 적어주신 부분은 잘 하셨습니다. 다른 방법은 아직 저도 모르겠어요.

junho15 committed:
refactor: EntryDetailViewController, ExpoEntriesListViewController, expoEntriesListTableView로 이름 수정, 공백 제거

EntryViewController -> EntryDetailViewController로 이름 수정
ExpoEntriesViewController -> ExpoEntriesListViewController로 이름 수정
ExpoEntriesTableView -> expoEntriesListTableView로 변수명 수정
공백 제거

yjjem committed:
refactor: Exposition 기능 분리 -> ExpositionManager

junho15 committed:
refactor: Extension 폴더 추가하여 분리

yjjem committed:
refactor: ExpoEntriesManager 생성하여 기능 분리

junho15 committed:
refactor: instantiateViewController(identifier:creator:)를 사용하도록 수정, configureEntryViews를 viewDidLoad()에서 호출하도록 수정

yjjem committed:
refactor: ExpoEntry 화면에 전달하는 데이터의 타입을 ExpoEntryData typealias로 수정

junho15 committed:
refactor: typealias 파일 분리, 네이밍 수정, import 정리

yjjem committed:
chore: 컨벤션 수정
